### PR TITLE
[WIP] Convert DATE to logical type

### PR DESCRIPTION
**Pull request by pramodsatya**

### CODE REVIEW

**Review by aditi-pandit — COMMENTED** [review]
Thanks Pramod for these changes. 

Have a single design question. If the usage of int64_t was unintentional then please change the code for this

File: velox/type/Type.h [hunk 1]
```diff
@@ -1093,6 +1076,54 @@ INTERVAL_DAY_TIME() {
   return IntervalDayTimeType::get();
 }
 
+/// Date can be represented as the number of days since epoch start using
+/// int64_t type.
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/type/Type.h]: Thanks for the suggestion, the use of int64_t was unintentional and it is changed to int32_t.

File: velox/type/Type.h [hunk 2]
```diff
@@ -1730,7 +1753,9 @@ template <>
 struct SimpleTypeTrait<Timestamp> : public TypeTraits<TypeKind::TIMESTAMP> {};
 
 template <>
-struct SimpleTypeTrait<Date> : public TypeTraits<TypeKind::DATE> {};
+struct SimpleTypeTrait<Date> : public SimpleTypeTrait<int64_t> {
```
Comment (aditi-pandit) [inline velox/type/Type.h:1756]: Why is it int64_t here ? My intuition was that it should be int32_t.

File: velox/type/tests/TypeTest.cpp [hunk 3]
```diff
@@ -127,8 +127,8 @@ TEST(TypeTest, date) {
   EXPECT_EQ(date->toString(), "DATE");
   EXPECT_EQ(date->size(), 0);
   EXPECT_THROW(date->childAt(0), std::invalid_argument);
-  EXPECT_EQ(date->kind(), TypeKind::DATE);
-  EXPECT_STREQ(date->kindName(), "DATE");
+  EXPECT_EQ(date->kind(), TypeKind::INTEGER);
```
Comment (aditi-pandit) [inline velox/type/tests/TypeTest.cpp:130]: We should add a test for date->name() to ensure that is DATE.

File: velox/type/Type.h [hunk 4]
```diff
@@ -1091,6 +1074,55 @@ INTERVAL_DAY_TIME() {
...
+
+  std::string toString() const override {
+    return name();
+  }
+
+  std::string toString(const int32_t days) const;
+
+  int32_t toDays(folly::StringPiece in) const;
+
+  int32_t toDays() const;
```
Comment (aditi-pandit) [inline velox/type/Type.h:1104]: What does this method do and why is it needed ?

File: velox/type/Type.h [hunk 5]
```diff
@@ -36,9 +36,9 @@
 #include "folly/CPortability.h"
 #include "velox/common/base/ClassName.h"
 #include "velox/common/serialization/Serializable.h"
-#include "velox/type/Date.h"
 #include "velox/type/StringView.h"
 #include "velox/type/Timestamp.h"
+#include "velox/type/TimestampConversion.h"
```
Comment (aditi-pandit) [inline velox/type/Type.h:41]: Why is this new header needed here ? It should work fine to just put it in the cpp file.

File: velox/connectors/hive/tests/HivePartitionFunctionTest.cpp [hunk 6]
```diff
@@ -261,11 +261,12 @@ TEST_F(HivePartitionFunctionTest, timestamp) {
 }
 
 TEST_F(HivePartitionFunctionTest, date) {
-  auto values = makeNullableFlatVector<Date>(
+  auto values = makeNullableFlatVector<int32_t>(
```
Comment (aditi-pandit) [inline velox/connectors/hive/tests/HivePartitionFunctionTest.cpp:264]: This test can be removed I think since partitioning by the Date struct is not needed anymore.

File: velox/dwio/dwrf/test/TestReader.cpp [hunk 7]
```diff
@@ -1538,11 +1538,12 @@ TEST(TestReader, testOrcReaderDate) {
...
   while (rowReader->next(1000, batch)) {
     auto rowVector = batch->as<RowVector>();
-    auto dates = rowVector->childAt(1)->as<SimpleVector<Date>>();
+    auto dates = rowVector->childAt(1)->as<SimpleVector<int32_t>>();
+    auto date = DATE();
 
     std::stringstream stream;
     stream << year << "-12-25";
-    EXPECT_EQ(stream.str(), dates->valueAt(0).toString());
+    EXPECT_EQ(stream.str(), date->toString(dates->valueAt(0)));
```
Comment (aditi-pandit) [inline velox/dwio/dwrf/test/TestReader.cpp:1546]: Can we just say DATE()->toString(...) here and avoid the temporary type?

File: velox/dwio/parquet/tests/ParquetReaderTestBase.h [hunk 8]
```diff
@@ -39,7 +39,7 @@ class ParquetReaderTestBase : public testing::Test {
   }
 
   static RowTypePtr dateSchema() {
-    return ROW({"date"}, {DATE()});
+    return ROW({"int"}, {INTEGER()});
```
Comment (aditi-pandit) [inline velox/dwio/parquet/tests/ParquetReaderTestBase.h:42]: Why is this change needed  ?

 static RowTypePtr dateSchema() {
    return ROW({"date"}, {DATE()}); seems reasonable.

File: velox/dwio/parquet/tests/duckdb_reader/ParquetReaderTest.cpp [hunk 9]
```diff
@@ -199,14 +193,14 @@ TEST_F(ParquetReaderTest, dateRead) {
   auto type = reader->typeWithId();
   EXPECT_EQ(type->size(), 1ULL);
   auto col0 = type->childAt(0);
-  EXPECT_EQ(col0->type->kind(), TypeKind::DATE);
+  EXPECT_EQ(col0->type->kind(), TypeKind::INTEGER);
```
Comment (aditi-pandit) [inline velox/dwio/parquet/tests/duckdb_reader/ParquetReaderTest.cpp:196]: You can check the type->name() here for date I think to check for DATE type in Parquet while reading.

File: velox/dwio/parquet/tests/reader/ParquetReaderTest.cpp [hunk 10]
```diff
@@ -92,7 +92,7 @@ TEST_F(ParquetReaderTest, parseDate) {
   auto type = reader.typeWithId();
   EXPECT_EQ(type->size(), 1ULL);
   auto col0 = type->childAt(0);
-  EXPECT_EQ(col0->type->kind(), TypeKind::DATE);
+  EXPECT_EQ(col0->type->kind(), TypeKind::INTEGER);
```
Comment (aditi-pandit) [inline velox/dwio/parquet/tests/reader/ParquetReaderTest.cpp:95]: we should check the type name here instead.

File: velox/dwio/type/fbhive/HiveTypeParser.cpp [hunk 11]
```diff
@@ -50,7 +50,6 @@ HiveTypeParser::HiveTypeParser() {
   setupMetadata<TokenType::Short, TypeKind::SMALLINT>("smallint");
   setupMetadata<TokenType::Integer, TypeKind::INTEGER>({"integer", "int"});
   setupMetadata<TokenType::Long, TypeKind::BIGINT>("bigint");
-  setupMetadata<TokenType::Date, TypeKind::DATE>("date");
```
Comment (aditi-pandit) [inline velox/dwio/type/fbhive/HiveTypeParser.cpp:53]: Doesn't seem reasonable to leave TokenType::Date left out. Please can you double check this.

File: velox/dwio/type/fbhive/HiveTypeParser.h [hunk 12]
```diff
@@ -34,7 +34,6 @@ enum class TokenType {
   Byte,
   Short,
   Integer,
-  Date,
```
Comment (aditi-pandit) [inline velox/dwio/type/fbhive/HiveTypeParser.h:37]: Why is Date removed here ?

File: velox/exec/tests/PrintPlanWithStatsTest.cpp [hunk 13]
```diff
@@ -261,8 +261,8 @@ TEST_F(PrintPlanWithStatsTest, partialAggregateWithTableScan) {
         {{"-- Aggregation\\[PARTIAL \\[c5\\] a0 := max\\(ROW\\[\"c0\"\\]\\), a1 := sum\\(ROW\\[\"c1\"\\]\\), a2 := sum\\(ROW\\[\"c2\"\\]\\), a3 := sum\\(ROW\\[\"c3\"\\]\\), a4 := sum\\(ROW\\[\"c4\"\\]\\)\\] -> c5:VARCHAR, a0:BIGINT, a1:BIGINT, a2:BIGINT, a3:DOUBLE, a4:DOUBLE"},
          {"   Output: .+, Cpu time: .+, Blocked wall time: .+, Peak memory: .+, Memory allocations: .+, Threads: 1"},
          {"      dataSourceLazyWallNanos\\s+sum: .+, count: 1, min: .+, max: .+"},
-         {"      hashtable.capacity\\s+sum: 1252, count: 1, min: 1252, max: 1252"},
-         {"      hashtable.numDistinct\\s+sum: 835, count: 1, min: 835, max: 835"},
+         {"      hashtable.capacity\\s+sum: 1273, count: 1, min: 1273, max: 1273"},
```
Comment (aditi-pandit) [inline velox/exec/tests/PrintPlanWithStatsTest.cpp:264]: Why did these changes happen ?

File: velox/functions/lib/RegistrationHelpers.h [hunk 14]
```diff
@@ -61,7 +60,7 @@ void registerNonSimdizableScalar(const std::vector<std::string>& aliases) {
   registerFunction<T, TReturn, Varbinary, Varbinary>(aliases);
   registerFunction<T, TReturn, bool, bool>(aliases);
   registerFunction<T, TReturn, Timestamp, Timestamp>(aliases);
-  registerFunction<T, TReturn, Date, Date>(aliases);
+  registerFunction<T, TReturn, int32_t, int32_t>(aliases);
```
Comment (aditi-pandit) [inline velox/functions/lib/RegistrationHelpers.h:63]: This line should be removed entirely. int32_t should be a simdizable scalar I think,

File: velox/functions/prestosql/GreatestLeast.cpp [hunk 15]
```diff
@@ -112,6 +112,15 @@ class ExtremeValueFunction : public exec::VectorFunction {
       exec::EvalCtx& context,
       VectorPtr& result) const override {
     switch (outputType.get()->kind()) {
+      case TypeKind::INTEGER:
+        if (isDateType(outputType)) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/GreatestLeast.cpp:116]: This is confusing.... I think if TypeKind::INTEGER should just call applyTyped... The check for Date type is not needed here.

File: velox/functions/prestosql/Sequence.cpp [hunk 16]
```diff
@@ -78,8 +78,10 @@ class SequenceFunction : public exec::VectorFunction {
     auto rawOffsets = offsets->asMutable<vector_size_t>();
 
     context.applyToSelectedNoThrow(rows, [&](auto row) {
-      auto start = toInt64(startVector->valueAt<T>(row));
-      auto stop = toInt64(stopVector->valueAt<T>(row));
+      T start1 = toInt64(startVector->valueAt<T>(row));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/Sequence.cpp:81]: Why are start1 and stop1 needed ?

File: velox/functions/prestosql/aggregates/tests/ArbitraryTest.cpp [hunk 17]
```diff
@@ -281,22 +281,15 @@ TEST_F(ArbitraryTest, date) {
       // Grouping key.
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/tests/ArbitraryTest.cpp:281]: This test is not needed anymore.

File: velox/functions/prestosql/aggregates/tests/ArbitraryTest.cpp [hunk 18]
```diff
@@ -281,22 +281,15 @@ TEST_F(ArbitraryTest, date) {
...
+          {125, 125, 126, 126, std::nullopt, std::nullopt, std::nullopt, 128},
+          DATE()),
       makeConstant<Timestamp>(std::nullopt, 8),
   });
 
   auto expectedResult = makeRowVector({
       makeFlatVector<int64_t>({1, 2, 3, 4}),
-      makeNullableFlatVector<Date>(
-          {Date(125), Date(126), std::nullopt, Date(128)}),
+      makeNullableFlatVector<int32_t>({125, 126, std::nullopt, 128}, DATE()),
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/tests/ArbitraryTest.cpp:292]: You can remove this vector also in the test.

File: velox/functions/prestosql/aggregates/tests/ChecksumAggregateTest.cpp [hunk 19]
```diff
@@ -138,9 +139,9 @@ TEST_F(ChecksumAggregateTest, reals) {
 }
 
 TEST_F(ChecksumAggregateTest, dates) {
-  assertSingleGroupChecksum<Date>({Date(0)}, "AAAAAAAAAAA=");
-  assertSingleGroupChecksum<Date>({Date(1)}, "vmaSXOnPGBc=");
-  assertSingleGroupChecksum<Date>({{}}, "h8rrhbF5N54=");
+  assertSingleGroupChecksum<int32_t>({0}, "AAAAAAAAAAA=", DATE());
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/tests/ChecksumAggregateTest.cpp:142]: This can be removed.

File: velox/functions/prestosql/aggregates/tests/HistogramTest.cpp [hunk 20]
```diff
@@ -166,16 +166,14 @@ TEST_F(HistogramTest, groupByDate) {
 
   auto vector1 = makeFlatVector<int32_t>(
       num, [](vector_size_t row) { return row % 3; }, nullEvery(4));
-  auto vector2 = makeFlatVector<Date>(
-      num, [](vector_size_t row) { return Date{row % 2}; }, nullEvery(5));
+  auto vector2 = makeFlatVector<int32_t>(
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/tests/HistogramTest.cpp:169]: Remove this test.

File: velox/functions/prestosql/aggregates/tests/HistogramTest.cpp [hunk 21]
```diff
@@ -234,11 +232,11 @@ TEST_F(HistogramTest, globalGroupByTimestamp) {
 
 TEST_F(HistogramTest, globalGroupByDate) {
   vector_size_t num = 10;
-  auto vector = makeFlatVector<Date>(
-      num, [](vector_size_t row) { return Date{row % 4}; }, nullEvery(7));
+  auto vector = makeFlatVector<int32_t>(
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/tests/HistogramTest.cpp:235]: Remove this test also.

File: velox/functions/prestosql/aggregates/tests/MaxSizeForStatsTest.cpp [hunk 22]
```diff
@@ -40,7 +40,7 @@ TEST_F(MaxSizeForStatsTest, nullValues) {
       makeNullableFlatVector<float>({std::nullopt, std::nullopt}),
       makeNullableFlatVector<double>({std::nullopt, std::nullopt}),
       makeNullableFlatVector<bool>({std::nullopt, std::nullopt}),
-      makeNullableFlatVector<Date>({std::nullopt, std::nullopt}),
+      makeNullableFlatVector<int32_t>({std::nullopt, std::nullopt}, DATE()),
```
Comment (aditi-pandit) [inline velox/functions/prestosql/aggregates/tests/MaxSizeForStatsTest.cpp:43]: Remove this column. Since Date is no longer supported here we needn't test. Same for other functions in this file.

**Comment by aditi-pandit** [comment]
@pramodsatya : Please can you rebase this PR.

File: velox/type/Type.h (hunk 2 above)
Comment (pramodsatya) [reply velox/type/Type.h]: Reverted, this was missed during the change, thanks for pointing it out.

File: velox/dwio/type/fbhive/HiveTypeParser.cpp (hunk 11 above)
Comment (pramodsatya) [reply velox/dwio/type/fbhive/HiveTypeParser.cpp:53]: Updated, thanks!

File: velox/dwio/type/fbhive/HiveTypeParser.h (hunk 12 above)
Comment (pramodsatya) [reply velox/dwio/type/fbhive/HiveTypeParser.h]: Reverted, thanks.

File: velox/functions/prestosql/GreatestLeast.cpp (hunk 15 above)
Comment (pramodsatya) [reply velox/functions/prestosql/GreatestLeast.cpp]: Got it, since the original check was for Date type only, this check was added. Reverted now, thanks!

File: velox/type/Type.h (hunk 5 above)
Comment (pramodsatya) [reply velox/type/Type.h]: Moved to the cpp file, thanks!
